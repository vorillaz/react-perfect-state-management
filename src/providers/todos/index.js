import React, {useState, useCallback} from 'react';
import getInitalState from './initialState';
import TodosContext from './context';

const StateProvider = ({children}) => {
  const [state, updateState] = useState(getInitalState());
  const update = useCallback(updateState, []);
  return (
    <TodosContext.Provider value={[state, update]}>
      {children}
    </TodosContext.Provider>
  );
};
export default StateProvider;
