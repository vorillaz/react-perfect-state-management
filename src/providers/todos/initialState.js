export default () => [
  {
    title: "Check the plants",
    completed: true,
    userId: 2
  },
  {
    title: "Go to the mall",
    completed: false,
    userId: 4
  },
  {
    title: "Write a song",
    completed: false,
    userId: 1
  }
];
