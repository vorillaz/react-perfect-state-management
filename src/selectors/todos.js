import { useCallback } from "react";
import { composeSelectors, getState, getUpdate } from "./shared";
import useContextSelector from "./useContextSelector";
import ContextTodos from "../providers/todos/context";

const getTodos = (todos = []) => todos;

export const useWithTodos = (Ctx = ContextTodos) => {
  const selector = useCallback(composeSelectors(getState, getTodos), []);
  return useContextSelector(Ctx, selector);
};

export const useWithAddTodos = (Ctx = ContextTodos) => {
  const selector = useCallback(composeSelectors(getUpdate), []);
  const update = useContextSelector(Ctx, selector);
  return todo => update(todos => [...todos, todo]);
};

export const useWithSetCompleteTodo = (Ctx = ContextTodos) => {
  const selector = useCallback(composeSelectors(getUpdate), []);
  const update = useContextSelector(Ctx, selector);
  return (indexToUpdate, completed) =>
    update(todos => {
      return [
        ...todos.slice(0, indexToUpdate),
        {
          ...todos[indexToUpdate],
          completed
        },
        ...todos.slice(indexToUpdate + 1)
      ];
    });
};
