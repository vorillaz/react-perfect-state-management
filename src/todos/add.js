import React, { useCallback, useState } from "react";
import { useWithAddTodos } from "../selectors/todos";
import { useWithUsers } from "../selectors/users";

export default () => {
  const [newItem, onNewItemUpdate] = useState("");
  const [userId, onSelectUserId] = useState(-1);
  const add = useWithAddTodos();
  const users = useWithUsers();

  const update = useCallback(e => {
    onNewItemUpdate(e.target.value);
  }, []);

  const onAdd = useCallback(
    e => {
      e.preventDefault();
      add({ title: newItem, userId, completed: false });
      onNewItemUpdate("");
      onSelectUserId(-1);
    },
    [add, newItem, userId]
  );
  const onUserPick = useCallback(e => {
    e.preventDefault();
    onSelectUserId(Number(e.target.value));
  }, []);

  return (
    <form className="w-full max-w-lg my-4" onSubmit={onAdd}>
      <div className="bg-white shadow-md rounded p-4 flex flex-wrap">
        <div className="w-full md:w mb-3 ">
          <input
            placeholder="✍ Add new todo "
            className="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline"
            type="text"
            value={newItem}
            onChange={update}
          />
        </div>
        <div className="w-full md:w-3/4 p-1 pl-0 pb-0 md:w">
          <div className="relative">
            <select
              className="block appearance-none w-full bg-gray-200 border border-gray-200 text-gray-700 py-3 px-4 pr-8 rounded leading-tight focus:outline-none focus:bg-white focus:border-gray-500"
              value={userId}
              onChange={onUserPick}
            >
              <option value="-1">Pick a user</option>
              {users.map(u => (
                <option key={`option-${u.id}`} value={u.id}>
                  {u.name}
                </option>
              ))}
            </select>
          </div>
        </div>
        <div className="w-full md:w-1/4 p-1 pb-0 md:w pr-0">
          <button
            className="w-full bg-green-500 hover:bg-green-900 text-white py-3 px-4 rounded  leading-tight"
            type="submit"
            disabled={newItem.trim() === ""}
          >
            Add new
          </button>
        </div>
      </div>
    </form>
  );
};
