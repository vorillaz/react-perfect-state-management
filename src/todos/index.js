import React from "react";
import { useWithTodos } from "../selectors/todos";
import Item from "./item";
import AddNew from "./add";

export default () => {
  const todos = useWithTodos();
  return (
    <div className="flex items-center justify-center h-screen mt-10">
      <div className="">
        <div className="w-full max-w-lg rounded overflow-hidden shadow-lg">
          <div className="px-6 py-4">
            {todos.map((todo, index) => (
              <Item todo={todo} index={index} key={`todo-${index}`} />
            ))}
          </div>
        </div>
        <AddNew />
      </div>
    </div>
  );
};
