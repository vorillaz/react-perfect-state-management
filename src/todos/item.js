import React, { useCallback } from "react";
import CheckboxIcon from "../svgs/checkboxIcon";
import { useWithSetCompleteTodo } from "../selectors/todos";
import { useWithUsersGetById } from "../selectors/users";

export default ({ todo, index }) => {
  const { title, completed, userId } = todo;
  const updateCompleted = useWithSetCompleteTodo();
  const user = useWithUsersGetById(userId);
  const onChangeUpdate = e => {
    updateCompleted(index, e.target.checked);
  };
  const onChange = useCallback(onChangeUpdate);
  return (
    <label className="todo">
      <input
        className="todo__state"
        checked={completed}
        onChange={onChange}
        type="checkbox"
      />
      <CheckboxIcon />
      <div className="todo__text text-xl font-medium flex items-center">
        {title}
        {user && (
          <span className="inline-block bg-gray-200 rounded-full ml-3 px-4 py-1 text-sm font-normal text-gray-700 mr-2">
            {user.name}
          </span>
        )}
      </div>
    </label>
  );
};
