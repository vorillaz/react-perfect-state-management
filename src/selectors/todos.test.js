import {renderHook} from '@testing-library/react-hooks';
import {createContext} from 'use-context-selector';
import {useWithTodos, useWithAddTodos, useWithSetCompleteTodo} from './todos';

const initialstate = [
  {
    title: 'Check the plants',
    completed: true,
    userId: 2
  },
  {
    title: 'Go to the mall',
    completed: false,
    userId: 4
  },
  {
    title: 'Write a song',
    completed: false,
    userId: 1
  }
];

it('useWithTodos', () => {
  const Ctx = createContext([initialstate]);
  const {result} = renderHook(() => useWithTodos(Ctx));
  expect(result.current).toMatchSnapshot();
});

xit('useWithAddTodos', () => {
  // Todo
});

xit('useWithSetCompleteTodo', () => {
  // Todo
});
