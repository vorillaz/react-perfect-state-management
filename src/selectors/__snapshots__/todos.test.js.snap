// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useWithTodos 1`] = `
Array [
  Object {
    "completed": true,
    "title": "Check the plants",
    "userId": 2,
  },
  Object {
    "completed": false,
    "title": "Go to the mall",
    "userId": 4,
  },
  Object {
    "completed": false,
    "title": "Write a song",
    "userId": 1,
  },
]
`;
